cilium/conntrack.cc

```diff
@@ -0,0 +1,282 @@
+    }
+    pair.first->second.reset(maps);
+  }
+  ENVOY_LOG(trace, "cilium.bpf_metadata: Opened IPv6 conntrack map {}", name);
+  return pair.first;
+}
+
+void CtMap::closeMaps(const std::shared_ptr<std::unordered_set<std::string>>& to_be_closed) {
+  std::lock_guard<std::mutex> guard(maps_mutex_);
+
+  for (const auto& map_name: *to_be_closed) {
+    const std::string& name = (map_name != "") ? map_name : "global";
+    auto ct4 = ct_maps4_.find(name);
+    if (ct4 != ct_maps4_.end()) {
+      ct_maps4_.erase(ct4);
+      ENVOY_LOG(trace, "cilium.bpf_metadata: Closed IPv4 conntrack map {}", name);
+    }
+    auto ct6 = ct_maps6_.find(name);
+    if (ct6 != ct_maps6_.end()) {
+      ct_maps6_.erase(ct6);
+      ENVOY_LOG(trace, "cilium.bpf_metadata: Closed IPv6 conntrack map {}", name);
+    }
+  }
+}
+
+CtMap::CtMap(const std::string &bpf_root) : bpf_root_(bpf_root) {
+  // Try to open the bpf maps from Cilium specific paths
+  openMap4("global");
+  openMap6("global");
+}
+
+uint32_t CtMap::lookupSrcIdentity(const std::string& map_name, const Network::Address::Ip* ip,
+				  const Network::Address::Ip* rip, bool ingress)
+{
+  const std::string& name = (map_name != "") ? map_name : "global";
+
+  ENVOY_LOG(debug, "cilium.bpf_metadata: Using conntrack map {}", name);
+
+  struct ipv4_ct_tuple key4 {};
+  struct ipv6_ct_tuple key6 {};
+  struct ct_entry value {};
+
+  if (ip->version() == Network::Address::IpVersion::v4 &&
+      rip->version() == Network::Address::IpVersion::v4) {
+    key4.daddr = ip->ipv4()->address();
+    key4.saddr = rip->ipv4()->address();
+    key4.sport = htons(rip->port());
+    key4.dport = htons(ip->port());
+    key4.nexthdr = 6; // TCP only for now
+    key4.flags = ingress ? TUPLE_F_IN : TUPLE_F_OUT; // also reversed
+
+    ENVOY_LOG(
+        trace,
+	"cilium.bpf_metadata: Looking up key: {:x}, {:x}, {:x}, {:x}, {:x}, {:x}",
+	ntohl(key4.daddr), ntohl(key4.saddr), ntohs(key4.dport), ntohs(key4.sport), key4.nexthdr,
+	key4.flags);
+  } else if (ip->version() == Network::Address::IpVersion::v6 &&
+	     rip->version() == Network::Address::IpVersion::v6) {
+    absl::uint128 daddr = ip->ipv6()->address();
+    absl::uint128 saddr = rip->ipv6()->address();
+    memcpy(&key6.daddr, &daddr, 16);
+    memcpy(&key6.saddr, &saddr, 16);
+    key6.sport = htons(rip->port());
+    key6.dport = htons(ip->port());
+    key6.nexthdr = 6; // TCP only for now
+    key6.flags = ingress ? TUPLE_F_IN : TUPLE_F_OUT;
+  } else {
+    ENVOY_LOG(
+        info,
+	"cilium.bpf_metadata: Address type mismatch: Source: {}, Dest: {}",
+	rip->addressAsString(), ip->addressAsString());
+    return 0;
+  }
+
+  if (ip->version() == Network::Address::IpVersion::v4) {
+    // Lock for the duration of the map lookup and conntrack lookup
+    std::lock_guard<std::mutex> guard(maps_mutex_);
+    auto it = ct_maps4_.find(name);
+    if (it == ct_maps4_.end()) {
+      it = openMap4(name);
+    }
+    if (it == ct_maps4_.end()) {
```

Comment:
nit: Seems like this if statement is the exact same check as three lines up, doesn't need to be a separate block.